### PR TITLE
feat: add custom component that generate logs and rotate wirh diff names

### PR DESCRIPTION
**Pull request by dgggIoT**
**Issue #, if available:**

**Description of changes:**
adding a custom component. It generate random messages and perform logging into the log file, which rotates with different name by using `timeAndSizeBasedRollingPolicy`. This will help OTF to perform UAT based on real log files. 
**Why is this change necessary:**

**How was this change tested:**

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public static void performLogging(Logger logger, int totalLogNumbers, long logWriteFreqSeconds, int logMsgSizeBytes) {
+        for (int i=0; i < totalLogNumbers; i++) {
+            random = Math.random();
+            StringBuilder sb = generateLogMessage(i, logMsgSizeBytes);
+            logger.info(sb.toString());
+            try {
+                TimeUnit.SECONDS.sleep(logWriteFreqSeconds);
+            } catch (InterruptedException e) {
```
Comment (MikeDombo) [inline uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:24]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)

**Comment by Nelsonochoam** [comment]
Build is failing

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
+/*
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:1]: Why that name `LogGeneratorDiffName.java`? You already deleted the previous file

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        fileSizeBytes = getFileSizeInBytes(fileSize, fileSizeUnit);
+
+        String currentPath = System.getProperty("user.dir");
+        Logger logger = prepareLogger(currentPath, logFileName, logFileExtension, fileSizeBytes);
+
+        logger.info("hello world !!!!!");
+
+        // starting logging message
+        performLogging(logger, totalLogNumbers, logWriteFreqSeconds, logMsgSizeBytes);
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:53]: Why not just add this method inside this class?

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        // print list of log files and results would be presented in local machine, it is easier to verify locally
+        // and manuaal. It is not used for the UAT verification.
+        verifyLoggingSucceed(currentPath, logFileName, logFileExtension);
+    }
+
+    private Logger prepareLogger(String currentPath, String fileName, String extension, int fileSizeBytes) {
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:60]:
```suggestion
    private Logger configureLogger(String currentPath, String fileName, String extension, int fileSizeBytes) {
```

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        logWriteFreqSeconds = Long.parseLong(args[4]);
+        logMsgSizeBytes = Integer.parseInt(args[5]);
+        totalLogNumbers = Integer.parseInt(args[6]);
+
+        fileSizeBytes = getFileSizeInBytes(fileSize, fileSizeUnit);
+
+        String currentPath = System.getProperty("user.dir");
+        Logger logger = prepareLogger(currentPath, logFileName, logFileExtension, fileSizeBytes);
+
+        logger.info("hello world !!!!!");
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:50]: Maybe we want to remove this line from being logged (I know you added it for testing)

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        // starting logging message
+        performLogging(logger, totalLogNumbers, logWriteFreqSeconds, logMsgSizeBytes);
+
+        // print list of log files and results would be presented in local machine, it is easier to verify locally
+        // and manuaal. It is not used for the UAT verification.
+        verifyLoggingSucceed(currentPath, logFileName, logFileExtension);
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:57]: I don't think we should add this. If we start printing additional stuff to stdout, the greengrass logger from the nucleus will intercept it and add them to a logging file I believe. Also you could do this by just listing the files on a directory this is getting written to on your machine

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                append(seq).
+                append(" " + namesArray[namesIndex]).
+                append(" " + verbsArray[verbsIndex]).
+                append(" " + nounsArray[nounsIndex]);
+
+        padLogMessage(sb, targetMsgSize);
+        return sb;
+    }
+
+    private static void padLogMessage(StringBuilder sb, int targetSize) {
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:48]: What is the intent behind padding the log messages? What sort of tests will it enable?. I guess it is so that we can test the disk space thread?.

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java (hunk 1 above)
Comment (Nelsonochoam) [reply uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:24]: We should not just print if someone interrupts then how will this exit upon interruption. We should maybe use `System.exit(0)`

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+public final class LoggingHelper {
+    private static double random;
+    public LoggingHelper() {
+    }
+
+    public static void performLogging(Logger logger, int totalLogNumbers, long logWriteFreqSeconds, int logMsgSizeBytes) {
+        for (int i=0; i < totalLogNumbers; i++) {
+            random = Math.random();
+            StringBuilder sb = generateLogMessage(i, logMsgSizeBytes);
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:20]: Why not return the logLine directly instead of the builder

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/UnitConvertHelper.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+package com.aws.greengrass.utils;
+
+import ch.qos.logback.core.util.FileSize;
+
+public final class UnitConvertHelper {
+    public UnitConvertHelper() {
+    }
+
+    public static int getFileSizeInBytes(double fileSize, String fileSizeUnit) {
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/utils/UnitConvertHelper.java:14]: Just put all this methods on the same class.

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java (hunk 2 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:1]: There will be another custom component that rotate files with the same name, basically using another logger rolling policy. I am trying to make them to be 2 components instead of 1.

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java (hunk 3 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:53]: Another component will also use the same logging but different logger.

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java (hunk 7 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:48]: My target logic of verifying data loss or duplicate in UAT scenario will be based on the number of log message we received. Thus, by fixing the size of each log message, I can always calculate the expected number of log msgs when given file size, then further compare it with what we receive on CW.

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java (hunk 8 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:20]: Following the style from GG.logging, it seems like they prefer the builder to be passed around.

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/UnitConvertHelper.java (hunk 9 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/java/com/aws/greengrass/utils/UnitConvertHelper.java:14]: This will be used by another component

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java (hunk 3 above)
Comment (Nelsonochoam) [reply uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:53]: Just use a base class and override it then

File: uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java (hunk 7 above)
Comment (Nelsonochoam) [reply uat/custom-components/src/main/java/com/aws/greengrass/utils/LoggingHelper.java:48]: I just think we should write logs like ` <seq> - <logger name> some random message`, then we just get all the logs from the log group on cloudwatch and assert there is not a missing sequence number for that component. That way we can know it uploaded everything and also keep it simple.

The only reason I see this would be useful is for testing the space management which is something we will get rid of in a later releasee

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java (hunk 6 above)
Comment (dgggIoT) [reply uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGeneratorDiffName.java:57]: Actually, this is my purpose. I want to check these info from the logging file that Nucleus writes to, because this file will appear in the /results in my local machine.

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java [hunk 10]
```diff
@@ -5,16 +5,200 @@
...
+    }
+
+    private static void performLogging(Logger logger, int totalLogNumbers, long logWriteFreqSeconds) {
+        for (int i=0; i < totalLogNumbers; i++) {
+            random = Math.random();
+            StringBuilder sb = generateLogMessage(i, logger.getName());
+            logger.info(sb.toString());
+            try {
+                TimeUnit.SECONDS.sleep(logWriteFreqSeconds);
+            } catch (InterruptedException e) {
```
Comment (MikeDombo) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java:111]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java [hunk 11]
```diff
@@ -5,16 +5,200 @@
...
+        System.out.println("listing files: " + pathnames.length);
+        for (String pathname : pathnames) {
+            System.out.println(pathname);
+        }
+    }
+
+    /*
+    This is only used to provide a way to read logs in file from local results, instead of SSH to the device.
+     */
+    private static void verifyFileIsWritten(File file) {
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java:171]: This would run on your local machine or the machine you launched the tests from. Why is this needed?

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java (hunk 10 above)
Comment (Nelsonochoam) [reply uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java:111]: Same comment I had left here, call `System.exit(0)` if interrupted

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java [hunk 12]
```diff
@@ -5,16 +5,141 @@
...
+    private String logFileName;
+    private String logFileExtension;
+    private double fileSize;
+    private String fileSizeUnit;
+    private double logWriteFreqSeconds;
+    private int logMsgSizeBytes;
+    private int fileSizeBytes;
+    private int totalLogNumbers;
+    // public only for unit testing
+    public String currentPath = System.getProperty("user.dir");
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java:38]: Let's not make public attributes just for testing, you could pass it on the accept method as an arg

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java [hunk 13]
```diff
@@ -5,16 +5,141 @@
...
+        appender.start();
+
+        // set logger
+        logger.addAppender(appender);
+        logger.setLevel(Level.INFO);
+
+        return logger;
+    }
+
+    public static void performLogging(Logger logger, int totalLogNumbers, double logWriteFreqSeconds) {
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java:96]: Why is this public?. If you really want to test it then make this private and the call `accept`. Test things from a consumers perspective, what the caller can actually do. That will make your tests impervious to changes allowing you to freely change your component. They will only change if you change the ordering of the params you use to call `accept`

File: uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java [hunk 14]
```diff
@@ -5,16 +5,141 @@
...
+                fileSizeBytes = (int) (fileSize * FileSize.KB_COEFFICIENT);
+                break;
+            case "MB":
+                fileSizeBytes = (int) (fileSize * FileSize.MB_COEFFICIENT);
+                break;
+            case "GB":
+                fileSizeBytes = (int) (fileSize * FileSize.GB_COEFFICIENT);
+                break;
+            default:
+                break;
```
Comment (Nelsonochoam) [inline uat/custom-components/src/main/java/com/aws/greengrass/artifacts/logGenerator.java:140]: Throw unsupported error?

File: uat/custom-components/src/test/java/com/aws/greengrass/LogGeneratorTest.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        assertTrue(Arrays.asList(pathnames).contains(activeFileName));
+        assertTrue(tempPath.resolve(activeFileName).toFile().length() > 0);
+    }
+
+    /*
+    if tempPath is empty string, then the log files would be generated in your local machine. '
+    This can be used for manually testing if logs are correctly written
+     */
+    @Test
+    void GIVEN_empty_targetFilePath_in_Paras_THEN_default_path_is_local()
```
Comment (Nelsonochoam) [inline uat/custom-components/src/test/java/com/aws/greengrass/LogGeneratorTest.java:48]:
```suggestion
    void GIVEN_empty_targetFilePath_THEN_default_path_is_local()
```

File: uat/custom-components/src/test/java/com/aws/greengrass/LogGeneratorTest.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    String logFileExtention = "log";
+    String logWriteFreqSeconds = "0.1";
+    String totalLogNumbers = "50";
+    String fileSizeBytes = "1024";
+    String fileSizeUnit = "KB";
+    String componentName = "com.aws.greengrass.artifacts.logGenerator";
+    String activeFileName = logFileName + "." + logFileExtention;
+
+    @Test
+    void GIVEN_request_THEN_log_file_Created()
```
Comment (Nelsonochoam) [inline uat/custom-components/src/test/java/com/aws/greengrass/LogGeneratorTest.java:29]:
```suggestion
    void GIVEN_request_THEN_log_file_created()
```

File: uat/custom-components/src/test/java/com/aws/greengrass/LogGeneratorTest.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+
+        String[] args = {logFileName, logFileExtention, fileSizeBytes, fileSizeUnit, logWriteFreqSeconds,
```
Comment (Nelsonochoam) [inline uat/custom-components/src/test/java/com/aws/greengrass/LogGeneratorTest.java:32]: nit: I would inline the variables here but it is ok to leave them as the class variables for now

**Review by Nelsonochoam — APPROVED** [review]
Approved with some comments

**Review by saranyailla — APPROVED** [review]
Change the PR and commit message to `test:add custom component that generate logs and rotate wirh diff names`
